### PR TITLE
tilelink: it is now legal to support Acquire for UNCACHED regions

### PR DESCRIPTION
**Pull request by terpstra**
These cases exist:
  GET_EFFECTS, PUT_EFFECTS, UNCACHEABLE && !supportsAcquire: MMIO
  UNCACHED && !supportsAcquire: speculation ok and may be cached
  UNCACHED && supportsAcquire: LLC/CacheCork applied (slave never probes)
  CACHED, TRACKED && supportsAcquire: slave might probe

### CODE REVIEW

File: src/main/scala/devices/tilelink/Error.scala [hunk 1]
```diff
@@ -40,35 +45,57 @@ class TLError(params: ErrorParams, beatBytes: Int = 4)(implicit p: Parameters) e
...
+
+    val a_last = edge.last(a)
+    val c_last = edge.last(c)
+    val da_last = edge.last(da)
+    val dc_last = edge.last(dc)
+
+    a.ready := (da.ready && da_last) || !a_last
+    da.valid := a.valid && a_last
+
+    val a_opcodes = Vec(AccessAck, AccessAck, AccessAckData, AccessAckData, AccessAckData, HintAck, Grant)
```
Comment (hcook) [inline src/main/scala/devices/tilelink/Error.scala]: this makes me nervous, should this be a utility func on `edge` or something?

File: src/main/scala/devices/tilelink/Error.scala [hunk 2]
```diff
@@ -40,35 +45,57 @@ class TLError(params: ErrorParams, beatBytes: Int = 4)(implicit p: Parameters) e
...
+    dc.bits.size   := c.bits.size
+    dc.bits.source := c.bits.source
+    dc.bits.sink   := UInt(0)
+    dc.bits.data   := UInt(0)
+    dc.bits.error  := Bool(true)
+
+    // Combine response channels
+    TLArbiter.lowest(edge, in.d, dc, da)
+
+    // We never probe or issue requests (UNCACHED)
```
Comment (hcook) [inline src/main/scala/devices/tilelink/Error.scala]: "requests" == "memory access ops on channel B"?

File: src/main/scala/devices/tilelink/Error.scala [hunk 3]
```diff
@@ -40,35 +45,57 @@ class TLError(params: ErrorParams, beatBytes: Int = 4)(implicit p: Parameters) e
...
     in.b.valid := Bool(false)
-    in.c.ready := Bool(true)
+
+    // Sink GrantAcks
     in.e.ready := Bool(true)
   }
 }
 
-trait HasPeripheryErrorSlave extends HasPeripheryBus {
+trait HasSystemErrorSlave extends HasPeripheryBus {
```
Comment (hcook) [inline src/main/scala/devices/tilelink/Error.scala]: you should change this to ` extends HasSystemBus`

File: src/main/scala/tilelink/CacheCork.scala [hunk 4]
```diff
@@ -61,6 +66,16 @@ class TLCacheCork(unsafe: Boolean = false)(implicit p: Parameters) extends LazyM
         a_a.bits.source := in.a.bits.source << 1 | UInt(1)
       }
 
+      // Upgrades are instantly successful
+      a_d.valid := in.a.valid && toD
+      a_d.bits.opcode := Grant
```
Comment (hcook) [inline src/main/scala/tilelink/CacheCork.scala]: is there a reason you aren't using the `TLEdgeIn.Grant` constructor here?

File: src/main/scala/devices/tilelink/Error.scala (hunk 1 above)
Comment (terpstra) [inline src/main/scala/devices/tilelink/Error.scala]: It's hard to generalize, because you don't know Grant or GrantData

**Pull request merged**